src/Http/Http.Extensions/gen/RequestDelegateGenerator.cs

```diff
@@ -69,11 +69,13 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
             codeWriter.EndBlockWithComma();
             codeWriter.WriteLine("(del, options, inferredMetadataResult) =>");
             codeWriter.StartBlock();
+            codeWriter.WriteLine(@"Debug.Assert(options != null, ""RequestDelegateFactoryOptions not found."");");
```

Comment:
I'm not super excited about this. Do we Assert in other generated code already?

It feels like we are doing something wrong. Can the user actually get into this situation? And if they can, do they know what to do to get out of it?

Reply:
Ha! @mitchdenny and I had a similar conversation about this yesterday.

> Do we Assert in other generated code already?

We introduced this assertion here in https://github.com/dotnet/aspnetcore/commit/b138bde0050c34d89f1360455103b59646d009d1. I'm not aware of the use `Debug.Assert` in other generated code (like the runtime's generators) but I'll attempt to justify why I think this is appropriate here.

> Can the user actually get into this situation?

Nope, not if they are using RDG.

The nullable parameters in the API signatures of `MetadataPopulator` and `RequestDelegateFactoryFunc` exists primarily because the `RequestDelegateFactory` has public APIs that allow uses to construct delegates and populate metadata  standalone _without_ going through our endpoint resolution logic (see [this](https://github.com/dotnet/aspnetcore/blob/9b6c17182d96d46ebfdc5e8359dc5b85245ce3b4/src/Http/Http.Extensions/src/RequestDelegateFactory.cs#L164) and [this](https://github.com/dotnet/aspnetcore/blob/9b6c17182d96d46ebfdc5e8359dc5b85245ce3b4/src/Http/Http.Extensions/src/RequestDelegateFactory.cs#L128). Internally, these APIs use fallbacks if null values are provided for the nullable parameters, as [seen here](https://github.com/dotnet/aspnetcore/blob/9b6c17182d96d46ebfdc5e8359dc5b85245ce3b4/src/Http/Http.Extensions/src/RequestDelegateFactory.cs#L263-L264).

With the generated code, the more specific `Map` overloads can only be called during endpoint resolution when RequestDelegateFactoryOptions are guaranteed to be set for route handlers ([ref](https://source.dot.net/#Microsoft.AspNetCore.Routing/RouteEndpointDataSource.cs,211), [ref](https://source.dot.net/#Microsoft.AspNetCore.Routing/RouteEndpointDataSource.cs,236)).

If there's every a scenario where null values are passed to the generated code, that means we did something wrong in our framework. There's not really anything the user can do to address that issue. IMO, the `Debug.Assert` here reflects that this is bookkeeping we're responsible for as framework authors, not a nullability requirement that the end-user violated.

Reply:
Would it make sense to just disable nullability checking in these files?

Reply:
> Would it make sense to just disable nullability checking in these files?

The nullability checks provide value when executing the requiredness checks that we have built in as part of parameter binding. Minimal APIs has a feature where in if you don't annotate a type as nullable, code-gen some logic that validates that it is provided (kind of like the `Required` attribute in `System.ComponentModel.DataAnnotations`). For example, the code below will bind the `age` value from the route parameter _but_ check that the route parameter has been provided before invoking the users handler.

```csharp
app.MapGet("/birthday/{age}", (int age) => ...)
```

Being able to leverage nullable flow analysis in generated code to ensure that we've done the appropriate requiredness checks is more valuable than not having it for the sake of avoiding two assertions, IMO.

Reply:
I would just `!` away the nullability checks on this one line:

```diff
-var serviceProvider = options?.ServiceProvider ?? options?.EndpointBuilder?.ApplicationServices;
+var serviceProvider = options!.ServiceProvider ?? options!.EndpointBuilder!.ApplicationServices;
```

If the user built for Release and somehow got into a `null` situation here, the behavior is the same. If they are in Debug and get asserts in here, they have no idea what to do either. Either way, it being a bug in our code.

Reply:
Yep. My preference towards using `Debug.Assert` vs. the null forgiving operator is that `Debug.Assert` is that we fail upfront in our debug builds if we're in an invalid state instead of waiting until the `options` are used. It makes more sense to be to alert about this inconsistency earlier.

Also, for users reading the generated code when in debug mode, I think the `Debug.Assert` communicates more clearly what is going on here than the null-forgiving operator (which personally gives me the ick when I see it in code that I don't own and don't understand why the developer made that guarantee).